### PR TITLE
Add support for watchdog timers.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/primitive_esp32.cc [hunk 1]
```diff
@@ -521,6 +529,51 @@ PRIMITIVE(memory_page_report) {
   return result;
 }
 
+PRIMITIVE(watchdog_init) {
+  ARGS(uint32, ms);
+
+  int watchdog = watchdog_timers.any();
+  if (watchdog == kInvalidWatchdogTimer) FAIL(OUT_OF_RANGE);
```
Comment (kasperl) [inline src/primitive_esp32.cc:536]: ALREADY_IN_USE?

File: src/primitive_esp32.cc [hunk 2]
```diff
@@ -521,6 +529,51 @@ PRIMITIVE(memory_page_report) {
...
+  if (err != ESP_OK) {
+    return Primitive::os_error(err, process);
+  }
+  return process->null_object();
+}
+
+PRIMITIVE(watchdog_deinit) {
+  esp_err_t err = esp_task_wdt_deinit();
+  if (err != ESP_OK) {
+    return Primitive::os_error(err, process);
```
Comment (kasperl) [inline src/primitive_esp32.cc:571]: If the watchdog is initialized, it almost feels like we need to make this FATAL and force a crash.

File: src/primitive_esp32.cc [hunk 3]
```diff
@@ -521,6 +529,51 @@ PRIMITIVE(memory_page_report) {
...
+    .idle_core_mask = 0,
+    .trigger_panic = true,
+  };
+  esp_err_t err = esp_task_wdt_init(&config);
+  if (err != ESP_OK) {
+    return Primitive::os_error(err, process);
+  }
+
+  SystemEventSource::instance()->run([&]() {
+    err = esp_task_wdt_add(null);  // Add the current thread.
```
Comment (kasperl) [inline src/primitive_esp32.cc:549]: I feel 'current' is overloaded here.

```suggestion
    err = esp_task_wdt_add(null);  // Add the SystemEventSource thread.
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/primitive_esp32.cc (hunk 1 above)
Comment (floitsch) [reply src/primitive_esp32.cc]: done.

File: src/primitive_esp32.cc (hunk 3 above)
Comment (floitsch) [reply src/primitive_esp32.cc]: done.

File: src/primitive_esp32.cc (hunk 2 above)
Comment (floitsch) [reply src/primitive_esp32.cc:571]: I'm not sure it's worth it: the only reason we can fail here is if a task is still registered, or if `deinit` was already called.
The bigger question (to me) is whether we should return the singleton even if `deinit` failed.

**Pull request merged**